lib/system/generic/CMakeLists.txt

```diff
@@ -17,10 +17,10 @@ collect (PROJECT_LIB_SOURCES irq.c)
 collect (PROJECT_LIB_SOURCES shmem.c)
 collect (PROJECT_LIB_SOURCES time.c)
 
-if ("${PROJECT_MACHINE}" STREQUAL "zynqmp_a53" OR
-    "${PROJECT_MACHINE}" STREQUAL "zynqmp_a72" OR
-    "${PROJECT_MACHINE}" STREQUAL "zynqmp_a78" OR
-    "${PROJECT_MACHINE}" STREQUAL "zynqmp_r5"  OR
+if ("${PROJECT_MACHINE}" STREQUAL "xlnx_a53" OR
+    "${PROJECT_MACHINE}" STREQUAL "xlnx_a72" OR
+    "${PROJECT_MACHINE}" STREQUAL "xlnx_a78" OR
+    "${PROJECT_MACHINE}" STREQUAL "xlnx_r5"  OR
```

Comment:
These machine names need to be chosen carefully.
So, Mulit-core platforms like zynqmp, has different configuration in same core. I can imagine three machine names in zynqmp: R5-lockstep-core, R5-split-core0 and R5-split-core1.

If we go that route, we might end up, with long list of MACHINE names, as future plaforms might have more such cores and configurations.

@arnopo , @wmamills any other approach in such case ?

Reply:
Or, we can introduce other cmake variables such as:

PROJECT_MACHINE=zynqmp, PROJECT_CORE=a53/R5, PROJECT_CORE_CONF=lockstep/split, PROJECT_CORE_NUMBER="0, 1 etc... ?

Reply:
Hi @tnmysh  from Xilinx-amd side, the core number and SOC for latest workflows (SDT) is inferred via header files. 

That is 'zynqmp' / 'versal' etc and core 0/ core1 are symbols in the BSP. That being said we can update BSP to output those two. lockstep/split is not always visible from tooling in Xilinx-amd so in many cases lockstep==rpu0 for the defaults.

if it is custom tooling then we can add separate handling for lockstep but again, rpu0 defaults should already be ok for lockstep unless you mean for lockstep it should be using all the available TCM banks for that cluster. please clarify what is needed for the lockstep case in default demos.

Reply:
though separately we would like to see if it is possible to remove the demos (can open separate PR / discussion for that)

Reply:
> That is 'zynqmp' / 'versal' etc and core 0/ core1 are symbols in the BSP.

In such case, I believe we shouldn't need xlnx_a53, xlnx_a72 both.
As zynqmp -> a53, versal -> a72. So, can we just have "xlnx" Instead?
How do we distinguish between cortex-a and cortex-r ? Headers also provide such defines?

Reply:
for cortex-a and cortex-r (and microblaze) we can distinguish from symbols. But if we do that for MB too then we will have to do something to account for the different logic in MB subsection.
can we have 'xlnx_a' , xlnx_r' and 'xlnx_mb' ?

Reply:
but @tnmysh  for downstream consumers (openamp) how will we determine the SOC/core etc in the CMake? do you want us to pick things up from CMake checksymbolexists ?

Reply:
Exactly, that is why I was suggesting to have more config variables, and those should be used across various software stacks i.e. apps, open-amp and libmetal.
My suggestion was for more generic case that can handle not xlnx platforms, but any multi-core platform.

May be we end up addressing it in different PR, but I am looking for suggestion that is scalable to other platforms too, not only xlnx platform.

Reply:
@tnmysh   adding adding machine name for SOC/Core combo is fair since we see that combination across multiple vendor/SOC's for remoteproc as well as in open source libs.

what that means is 'versal_a72' is a common-sense convention. it has the drawback of being redundant with headers but scales for other SOC's but is sufficient for many cases. Core or vendor specific configurations (lockstep for rpu) can be captured in header symbols.

Reply:
PROJECT_MACHINE is not a generic variable and seems only used by AMD.

The objective here is to include `xlnx` subdirectory, right ?

What about a directory hierarchy `vendor/platform/core` or `vendor/core`?

a `PROJECT_VENDOR` ( or another variable) could define the vendor sub-directory and then  in vendor directory the core can be differentiated, if needed, by `PROJECT_MACHINE` or `PROJECT_TARGET`

Reply:
Hi @arnopo  thanks for the comment here.
so if I understand your suggestion then the libmetal usage would be as follows:

- add new cmake var PROJECT_VENDOR
- in libmetal lib ports we can just keep the vendor (each vendor can then organize that as needed, xilinx-amd would just keep the xlnx subdir)

in cmake lib simply validate PROJECT_VENDOR and then in the vendor-specific CMake there can be additional validation of the  PROJECT_MACHINE/PROJECT_TARGET as needed

@tnmysh  @arnopo  are you OK with that proposal?

Reply:
This triggers a discussion on how to harmonize the folders in the different repositories to add more platform support. Let's add this to the agenda for the next RP call.